src/NuGet.Clients/NuGet.SolutionRestoreManager/InfoBarService.cs

```diff
@@ -0,0 +1,164 @@
+
+            if (_closed || _visible)
+            {
+                return;
+            }
+
+            IVsInfoBarHost? infoBarHost;
+            try
+            {
+                IVsUIShell? uiShell = await _asyncServiceProvider.GetServiceAsync<SVsUIShell, IVsUIShell>(throwOnFailure: false);
+                if (uiShell == null)
+                {
+                    throw new NullReferenceException(nameof(uiShell));
+                }
+
+                if (ErrorHandler.Failed(uiShell.FindToolWindow((uint)__VSFINDTOOLWIN.FTW_fFindFirst, VSConstants.StandardToolWindows.SolutionExplorer, out var windowFrame)))
+                {
+                    throw new NullReferenceException(nameof(windowFrame));
+                }
+
+                object tempObject;
+                if (ErrorHandler.Failed(windowFrame.GetProperty((int)__VSFPROPID7.VSFPROPID_InfoBarHost, out tempObject)))
+                {
+                    throw new NullReferenceException(nameof(tempObject));
+                }
+
+                infoBarHost = (IVsInfoBarHost)tempObject;
+            }
+            catch (Exception ex)
+            {
+                await TelemetryUtility.PostFaultAsync(ex, nameof(InfoBarService));
+                return;
+            }
+
+            try
+            {
+                IVsInfoBarUIFactory? infoBarFactory = await _asyncServiceProvider.GetServiceAsync<SVsInfoBarUIFactory, IVsInfoBarUIFactory>(throwOnFailure: false);
+                if (infoBarFactory == null)
+                {
+                    NullReferenceException exception = new NullReferenceException(nameof(infoBarFactory));
+                    await TelemetryUtility.PostFaultAsync(exception, nameof(InfoBarService));
+                    return;
+                }
+
+                InfoBarModel infoBarModel = GetInfoBarModel();
+
+                _infoBarUIElement = infoBarFactory.CreateInfoBar(infoBarModel);
+                _infoBarUIElement.Advise(this, out uint cookie);
+                _eventCookie = cookie;
+
+                infoBarHost.AddInfoBar(_infoBarUIElement);
+
+                _visible = true;
+            }
+            catch
+            {
+                return;
+            }
+        }
+
+        public void OnClosed(IVsInfoBarUIElement infoBarUIElement)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            if (_eventCookie.HasValue)
+            {
+                infoBarUIElement?.Unadvise(_eventCookie.Value);
+                infoBarUIElement?.Close();
+                _eventCookie = null;
+            }
+
+            _visible = false;
+
+            if (!_closeFromHide)
+            {
+                _closed = true;
+            }
+        }
+
+        public void OnActionItemClicked(IVsInfoBarUIElement infoBarUIElement, IVsInfoBarActionItem actionItem)
+        {
+            PackageManagerUIStarter?.Value.PMUIStarter();
```

Comment:
I'm trying to think of how we would not be on the UI thread... is it guaranteed here?

Reply:
Adding a ThreadHelper.ThrowIfNotOnUIThread(); to be safe